pykrige/core.py

```diff
@@ -194,29 +191,70 @@ def _adjust_for_anisotropy(X, center, scaling, angle):
     return X_adj
 
 
-def initialize_variogram_model(x, y, z, variogram_model, variogram_model_parameters,
-                               variogram_function, nlags, weight, coordinates_type):
-    """Initializes the variogram model for kriging according
-    to user specifications or to defaults"""
+def initialize_variogram_model(X, y, variogram_model, variogram_model_parameters, variogram_function,
+                               nlags, weight, coordinates_type):
+    """Initializes the variogram model for kriging. If user does not specify parameters,
+    calls automatic variogram estimation routine.
 
-    x1, x2 = np.meshgrid(x, x, sparse=True)
-    y1, y2 = np.meshgrid(y, y, sparse=True)
-    z1, z2 = np.meshgrid(z, z, sparse=True)
-    dz = z1 - z2
+    Parameters
+    ----------
+    X: ndarray
+        float array [n_samples, n_dim], the input array of coordinates
+    y: ndarray
+        float array [n_samples], the input array of values to be kriged
+    variogram_model: str/unicode
+        user-specified variogram model to use
+    variogram_model_parameters: list
+        user-specified parameters for variogram model
+    variogram_function: callable
+        function that will be called to evaluate variogram model
+        (only used if user does not specify variogram model parameters)
+    nlags: int
+        integer scalar, number of bins into which to group inter-point distances
+    weight: bool
+        boolean flag that indicates whether the semivariances at smaller lags
+        should be weighted more heavily in the automatic variogram estimation
+    coordinates_type: str/unicode
+        type of coordinates in X array, can be 'euclidean' for standard rectangular coordinates
+        or 'geographic' if the coordinates are lat/lon
+
+    Returns
+    -------
+    lags: ndarray
+        float array [nlags], distance values for bins into which the semivariances were grouped
+    semivariance: ndarray
+        float array [nlags], averaged semivariance for each bin
+    variogram_model_parameters: list
+        parameters for the variogram model, either returned unaffected if the user specified them or
+        returned from the automatic variogram estimation routine
+    """
 
+    # distance calculation for rectangular coords now leverages scipy.spatial.distance's pdist function,
+    # which gives pairwise distances in a condensed distance vector (distance matrix flattened to a vector)
+    # to calculate semivariances... trick scipy.spatial.distance's pdist function into thinking that the measurements
+    # are in a 2D space by stacking with an array of zeros
     if coordinates_type == 'euclidean':
-        dx = x1 - x2
-        dy = y1 - y2
-        d = np.sqrt(dx**2 + dy**2)
+        d = pdist(X, metric='euclidean')
+        y_temp = np.vstack((y, np.zeros(y.size))).T
+        g = 0.5 * pdist(y_temp, metric='sqeuclidean')
```

Comment:
Good suggestion, I'll go with ```y[:, None]```
The full square matrix here previously was just kind of a stepping stone to get to the vector of unique distances, so using ```pdist``` helps to clean things up I think.